addOns/requester/src/main/java/org/zaproxy/addon/requester/internal/HttpPanelSender.java

```diff
@@ -255,6 +265,29 @@ private JToggleButton getButtonHostHeader() {
         return hostHeader;
     }
 
+    private JButton getButtonLowerCaseHeaderName() {
+        if (lowerCaseHeaderNames == null) {
+            lowerCaseHeaderNames =
+                    new JButton(ExtensionRequester.createIcon("lowercase-header-button.png"));
+            lowerCaseHeaderNames.setToolTipText(
+                    Constant.messages.getString(
+                            "requester.httpsender.button.lowerCaseHeadersName"));
+            lowerCaseHeaderNames.addActionListener(e -> lowerCaseHeaderNames());
+        }
+        return lowerCaseHeaderNames;
+    }
+
+    private void lowerCaseHeaderNames() {
+        customHttpPanelRequest.saveData();
+        HttpRequestHeader httpRequestHeader =
+                ((HttpMessage) customHttpPanelRequest.getMessage()).getRequestHeader();
+        for (HttpHeaderField field : httpRequestHeader.getHeaders()) {
+            httpRequestHeader.setHeader(field.getName(), null);
+            httpRequestHeader.addHeader(field.getName().toLowerCase(Locale.ROOT), field.getValue());
+        }
```

Comment:
I gave it a try and it appears to be working as expected. Would you mind sharing some additional information?

Reply:
Try with two fields with the same name.

Reply:
The least amount of change/deviation from original. The user/code may have specifically wanted/needed two. So to answer your question the message should still have two lowercase headers with the same name.

Reply:
ok done

Reply:
Unit test should still be added though.

Reply:
I had a quick look at that and couldn't sort out the mocking. In particular `CustomHttpPanelRequest` seems to be a weird cases that mockito claims it or it's supers aren't reachable.

Reply:
Something like this?

```diff
$ git diff HEAD
diff --git a/addOns/requester/src/main/java/org/zaproxy/addon/requester/internal/HttpPanelSender.java b/addOns/requester/src/main/java/o
rg/zaproxy/addon/requester/internal/HttpPanelSender.java
index 77e343974..97e993119 100644
--- a/addOns/requester/src/main/java/org/zaproxy/addon/requester/internal/HttpPanelSender.java
+++ b/addOns/requester/src/main/java/org/zaproxy/addon/requester/internal/HttpPanelSender.java
@@ -292,6 +292,17 @@ public class HttpPanelSender {
         customHttpPanelRequest.updateContent();
     }

+    static void testLowerCaseHeaderNames(HttpMessage msg) {
+        HttpRequestHeader httpRequestHeader = msg.getRequestHeader();
+        List<HttpHeaderField> fields = httpRequestHeader.getHeaders();
+        for (HttpHeaderField field : fields) {
+            httpRequestHeader.setHeader(field.getName(), null);
+        }
+        for (HttpHeaderField field : fields) {
+            httpRequestHeader.addHeader(field.getName().toLowerCase(Locale.ROOT), field.getValue());
+        }
+    }
+
     /**
      * A {@link HttpRedirectionValidator} that enforces the {@link
      * org.parosproxy.paros.control.Control.Mode Mode} when validating the {@code URI} of
diff --git a/addOns/requester/src/test/java/org/zaproxy/addon/requester/internal/HttpPanelSenderUnitTest.java b/addOns/requester/src/tes
t/java/org/zaproxy/addon/requester/internal/HttpPanelSenderUnitTest.java
new file mode 100644
index 000000000..0c6d5618a
--- /dev/null
+++ b/addOns/requester/src/test/java/org/zaproxy/addon/requester/internal/HttpPanelSenderUnitTest.java
@@ -0,0 +1,44 @@
+package org.zaproxy.addon.requester.internal;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+
+import java.io.IOException;
+
+import org.apache.commons.httpclient.URI;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.parosproxy.paros.network.HttpHeaderField;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.network.HttpRequestHeader;
+
+class HttpPanelSenderUnitTest {
+
+    private static final String URI = "http://example.com";
+    private HttpMessage msg;
+
+    @BeforeEach
+    void init() throws IOException {
+        HttpRequestHeader requestHeader = new HttpRequestHeader();
+        requestHeader.setURI(new URI(URI, false));
+
+        msg = new HttpMessage();
+        msg.setRequestHeader(requestHeader);
+    }
+
+    @Test
+    void shouldLowercaseheaderNames() {
+        // Given
+        msg.getRequestHeader().addHeader("X-Foo", "bar");
+        msg.getRequestHeader().addHeader("X-Client", "Foo-bar 1.1.0");
+        // When
+        HttpPanelSender.testLowerCaseHeaderNames(msg);
+        // Then
+        assertThat(msg.getRequestHeader().getHeaders().size(), is(equalTo(2)));
+        assertThat(msg.getRequestHeader().getHeaders(),
+                containsInAnyOrder(new HttpHeaderField("x-foo", "bar"),
+                        new HttpHeaderField("x-client", "Foo-bar 1.1.0")));
+    }
+}
```

Reply:
I'm learning about unit testing but don't have much experience. Could you recommend some resources or sample tests for reference? Thanks!

Reply:
Assuming @thc202 or someone else agrees you can apply the above diff/patch then simply add a test that checks the duplicate header situation.

You can also definitely read through tests in any of the add-ons to get ideas/guideance.

Reply:
The new method would be called by the existing though, but yeah, that looks good.

Reply:
I am wondering if I need to create a new test method like testLowerCaseHeaderNames in the HttpPanelSender.java file to test the lowerCaseHeaderNames function. Additionally, I noticed that the Requester class appears to have very few unit tests, with only RequesterParamUnitTest.java being present. why is it so ?

Reply:
Yes as outlined/agreed above.

It’s short on tests because it’s old code.

Reply:
> The new method would be called by the existing though, but yeah, that looks good.

@amitpanwar789 this still needs to be done, maybe instead of test* call it set* and use it from the button method (then the code is t duplicated). The method probably also doesn't need "public" visibility, try just "static void".

Reply:
but we need to make it public otherwise i am getting error in HttpPanelSenderUnitTest file "'setLowerCaseHeaderNames(org.parosproxy.paros.network.HttpMessage)' is not public in 'org.zaproxy.addon.requester.internal.HttpPanelSender'. Cannot be accessed from outside package"

```
    public void lowerCaseHeaderNames() {
        customHttpPanelRequest.saveData();
        HttpMessage msg = ((HttpMessage) customHttpPanelRequest.getMessage());
        setLowerCaseHeaderNames(msg);
        customHttpPanelRequest.updateContent();
    }

    static void setLowerCaseHeaderNames(HttpMessage msg) {
        HttpRequestHeader httpRequestHeader = msg.getRequestHeader();
        List<HttpHeaderField> fields = httpRequestHeader.getHeaders();
        for (HttpHeaderField field : fields) {
            httpRequestHeader.setHeader(field.getName(), null);
        }
        for (HttpHeaderField field : fields) {
            httpRequestHeader.addHeader(field.getName().toLowerCase(Locale.ROOT), field.getValue());
        }
    }

```

Reply:
Okay that's fine, I wasn't totally sure it would work that's why I said "try" 😁

Reply:
It does not need to be public, the problem is that the test class is not in the same package as the class being tested.

Reply:
Good catch. @amitpanwar789 if you add the unit test class to `intetnal` (a new folder) then it should work without being public.

Reply:
ok